src/main/java/com/minecolonies/coremod/entity/ai/citizen/sifter/EntityAIWorkSifter.java

```diff
@@ -99,27 +103,36 @@ protected IAIState sift()
             return INVENTORY_FULL;
         }
 
-        currentRecipeStorage = sifterBuilding.getFirstFullFillableRecipe(item -> ItemStackUtils.isEmpty(item), 1, false);
+        currentRecipeStorage = sifterBuilding.getFirstFullFillableRecipe(ItemStackUtils::isEmpty, 1, false);
 
         if (currentRecipeStorage == null)
         {
+            if (!ItemStackUtils.isEmpty(worker.getHeldItemMainhand()))
+            {
+                worker.setHeldItem(Hand.MAIN_HAND, ItemStack.EMPTY);
+            }
+            if (!ItemStackUtils.isEmpty(worker.getHeldItemOffhand()))
+            {
+                worker.setHeldItem(Hand.OFF_HAND, ItemStack.EMPTY);
+            }
+
             progress = 0;
             return START_WORKING;
         }
 
-        ItemStack meshItem = ItemStack.EMPTY;
-        for (final IItemHandler handler : InventoryUtils.getItemHandlersFromProvider(sifterBuilding))
+        final ItemStack meshItem = currentRecipeStorage.getCraftingTools().get(0);
+        final ItemStack inputItem = currentRecipeStorage.getCleanedInput().stream()
+                .map(ItemStorage::getItemStack)
+                .filter(item -> !ItemStackUtils.compareItemStacksIgnoreStackSize(item, meshItem, false, true))
+                .findFirst().orElse(ItemStack.EMPTY);
+
+        if (!inputItem.isEmpty() && (ItemStackUtils.isEmpty(worker.getHeldItemMainhand()) || ItemStackUtils.compareItemStacksIgnoreStackSize(worker.getHeldItemMainhand(), inputItem)))
         {
-            final int foundSlot = InventoryUtils.findFirstSlotInItemHandlerNotEmptyWith(handler, item -> ItemStackUtils.compareItemStackListIgnoreStackSize(currentRecipeStorage.getCraftingTools(), item, false, true));
-            if (foundSlot > -1)
-            {
-                meshItem = handler.getStackInSlot(foundSlot).copy();
-            }
+            worker.setHeldItem(Hand.MAIN_HAND, inputItem);
         }
-
-        if(!meshItem.isEmpty() && (ItemStackUtils.isEmpty(worker.getHeldItemMainhand()) || ItemStackUtils.compareItemStacksIgnoreStackSize(worker.getHeldItemMainhand(), meshItem, false, true)))
+        if (!meshItem.isEmpty() && (ItemStackUtils.isEmpty(worker.getHeldItemOffhand()) || ItemStackUtils.compareItemStacksIgnoreStackSize(worker.getHeldItemOffhand(), meshItem, false, true)))
         {
-            worker.setHeldItem(Hand.MAIN_HAND, meshItem);
+            worker.setHeldItem(Hand.OFF_HAND, meshItem);
         }
```

Comment:
This makes the sifter hold the mesh in their offhand and the block being sifted in their main hand (and still swinging their main hand), which I think leads to a better animation.  (And the individual loot tables no longer care which hand the mesh is in, so that's not a compatibility issue.)

I've also removed the inventory search since the held items (as far as I can tell) are just for display purposes, so it doesn't matter whether it finds the "real" damaged mesh or the pristine mesh from the recipe.

Reply:
I'm not sure if the "don't set if already the correct item" checks are strictly necessary (I would think that would be checked inside the function), but I kept them since they were used in the original code.

Reply:
The reason for the inventory search (and no, it didn't work yet) was to support possible enchantments on the meshes, like Unbreaking, where the pristine one in the recipe didn't have that enchantment.  The "don't set" checks are because there aren't a lot of checks. I found this out the hard way working on the smeltercrafter, where I set it everytime, and the block flickered in their hand.

Reply:
True, an enchanted mesh would be a visible difference -- but then an enchanted mesh won't actually work with the code as written (and as before) anyway, because that's not an NBT match to the recipe.  Worry about that when it's actually implemented, probably. 😄

Having said that, I don't think it's too immersion-breaking to show the unenchanted mesh in their hands, it's mostly just to indicate which class of mesh they're using to make sure that the recipes are in the correct order in the hut.  (On that note, something I noticed both before and after these changes is that the recipes don't seem to be in any particular order -- notably string meshes appear both above and below flint meshes.  I thought you said that you had wanted them to be sorted with better meshes first?  I didn't investigate that, but it doesn't appear to be happening.)